pkg/cluster/cache/readiness.go

```diff
@@ -0,0 +1,260 @@
+package cache
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+
+	harbormetav1 "github.com/goharbor/harbor-operator/apis/meta/v1alpha1"
+
+	rediscli "github.com/go-redis/redis"
+	"github.com/goharbor/harbor-operator/apis/goharbor.io/v1alpha2"
+	"github.com/goharbor/harbor-operator/pkg/lcm"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	labels1 "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	cachePropertiesKey = "cache"
+)
+
+// Readiness reconcile will check Redis sentinel cluster if that has available.
+// It does:
+// - create redis connection pool
+// - ping redis server
+// - return redis properties if redis has available
+func (rc *RedisController) Readiness(cluster *v1alpha2.HarborCluster) (*lcm.CRStatus, error) {
+	var (
+		client *rediscli.Client
+		err    error
+	)
+
+	client, err = rc.GetInClusterRedisInfo(cluster)
+	if err != nil {
+		rc.Log.Error(err, "Fail to create redis client.",
+			"namespace", cluster.Namespace, "name", cluster.Name)
+		return cacheNotReadyStatus(ErrorGetRedisClient, err.Error()), err
+	}
+
+	defer client.Close()
+
+	if err := client.Ping().Err(); err != nil {
+		rc.Log.Error(err, "Fail to check Redis.",
+			"namespace", cluster.Namespace, "name", cluster.Name)
+		return cacheNotReadyStatus(ErrorCheckRedisHealth, err.Error()), err
+	}
+
+	rc.Log.Info("Redis already ready.",
+		"namespace", cluster.Namespace, "name", cluster.Name)
+
+	properties := lcm.Properties{}
+	properties.Add(cachePropertiesKey, rc.generateRedisSpec())
+	return cacheReadyStatus(&properties), nil
+}
+
+func (rc *RedisController) generateRedisSpec() *v1alpha2.ExternalRedisSpec {
+	return &v1alpha2.ExternalRedisSpec{
+		RedisHostSpec: harbormetav1.RedisHostSpec{
+			Host: rc.ResourceManager.GetService().Name,
```

Comment:
Why directly call a new method `GetServiceName()`? So needn't generate the whole Service again.